### PR TITLE
Create ML challenge and set day of problem

### PR DESCRIPTION
**Pull request by alkeshrails**
Create ML challenge with days limit : N days of ML

https://docs.google.com/document/d/1krhbByFZxyUeqfC9IZsK4w2BA28DXR4mewmqSrSjyDc/edit#heading=h.vy9pu42nawu8

Run rack to create record : `rake create_daily_practice_goal:seed_record`
run to create ActivityPoint record `rake activity_point:create_seed_activity`

### CODE REVIEW

File: app/services/calculate_meta_leaderboard_service.rb [hunk 1]
```diff
@@ -53,6 +63,7 @@ def participant_values(rank, participant, scores)
       submission_id: scores['submission_id'],
       seq: rank,
       meta: scores['rank'],
+      ml: scores['rank'],
```
Comment (skbly7) [inline app/services/calculate_meta_leaderboard_service.rb:66]: Remove this extra field, not needed

File: app/services/ml_challenge/award_point_service.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
+module MlChallenge
```
Comment (piotrekpasciak) [inline app/services/ml_challenge/award_point_service.rb:1]: This class doesn't seem to have test coverage. Please add it.

File: app/services/challenge_rounds/create_leaderboards_service.rb [hunk 3]
```diff
@@ -1,15 +1,19 @@
 module ChallengeRounds
   class CreateLeaderboardsService < ::BaseService
```
Comment (piotrekpasciak) [inline app/services/challenge_rounds/create_leaderboards_service.rb:2]: Please add test coverage for new new logic in this class.

File: app/models/participant.rb [hunk 4]
```diff
@@ -109,6 +109,8 @@ class Participant < ApplicationRecord
   has_many :newsletter_emails, class_name: 'NewsletterEmail', dependent: :destroy
   has_many :notifications, dependent: :destroy
   has_many :team_members, dependent: :destroy
+  has_many :participant_ml_challenge_goals
```
Comment (piotrekpasciak) [inline app/models/participant.rb:112]: Please define dependent option for this line and one below.

File: app/models/leaderboard/base_leaderboard.rb [hunk 5]
```diff
@@ -3,6 +3,7 @@ class BaseLeaderboard < ApplicationRecord
 
   belongs_to :challenge
   belongs_to :meta_challenge, optional: true, class_name: 'Challenge'
+  belongs_to :ml_challenge, optional: true, class_name: 'Challenge'
```
Comment (piotrekpasciak) [inline app/models/leaderboard/base_leaderboard.rb:6]: What about  Leaderboard and OngoingLeaderboard models? Missing relation may cause issues there.

File: app/models/daily_practice_goal.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class DailyPracticeGoal < ApplicationRecord
+  validates :title, :points, presence: true
+
+  has_many :participant_ml_challenge_goals
```
Comment (piotrekpasciak) [inline app/models/daily_practice_goal.rb:4]: Please define dependent option.

File: app/models/challenge.rb [hunk 7]
```diff
@@ -65,6 +65,8 @@ class Challenge < ApplicationRecord
   has_many :categories, through: :category_challenges
   has_many :newsletter_emails, class_name: 'NewsletterEmail'
   has_many :notifications, class_name: 'Notification'
+  has_many :participant_ml_challenge_goals
```
Comment (piotrekpasciak) [inline app/models/challenge.rb:68]: Please define `dependent` option for this relation.

File: app/models/activity_point.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+class ActivityPoint < ApplicationRecord
+  validates :activity_key, uniqueness: true, presence: true
+  validates :point, presence: true
+
+  has_many :ml_activity_points
```
Comment (piotrekpasciak) [inline app/models/activity_point.rb:5]: Please define dependent option for this relation.

File: app/controllers/daily_practice_goals_controller.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+class DailyPracticeGoalsController < ApplicationController
+  before_action :authenticate_participant!
+  before_action :set_challenge, only: [:index]
+
+  def index
```
Comment (piotrekpasciak) [inline app/controllers/daily_practice_goals_controller.rb:5]: Please add simple feature spec for this page.

File: db/migrate/20200707145533_create_ml_activity_points.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateMlActivityPoints < ActiveRecord::Migration[5.2]
+  def change
+    create_table :ml_activity_points do |t|
+      t.integer :participant_id
```
[the remark this answers is not in the document]
Comment (piotrekpasciak) [reply db/migrate/20200707145533_create_ml_activity_points.rb]: Froeign keys are most important to keep the integrity of data, but can't find it schema.rb file at all.

File: db/migrate/20200709123607_add_ml_challenge_id_to_submission.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+class AddMlChallengeIdToSubmission < ActiveRecord::Migration[5.2]
+  def change
+    add_column :submissions, :ml_challenge_id, :integer
```
[the remark this answers is not in the document]
Comment (piotrekpasciak) [reply db/migrate/20200709123607_add_ml_challenge_id_to_submission.rb:3]: Can't find index, nor foreign_key.

File: app/models/participant.rb (hunk 4 above)
Comment (piotrekpasciak) [reply app/models/participant.rb]: Line below is still missing depedent option.

File: app/models/activity_point.rb (hunk 8 above)
Comment (piotrekpasciak) [reply app/models/activity_point.rb]: Depedent option is still missing

File: app/models/leaderboard/base_leaderboard.rb (hunk 5 above)
Comment (piotrekpasciak) [reply app/models/leaderboard/base_leaderboard.rb:6]: I still don't see this change in ongoing_leaderboard.rb

File: app/services/calculate_meta_leaderboard_service.rb (hunk 1 above)
Comment (alkeshrails) [reply app/services/calculate_meta_leaderboard_service.rb]: Done

File: app/models/daily_practice_goal.rb (hunk 6 above)
Comment (alkeshrails) [reply app/models/daily_practice_goal.rb]: Done

File: app/models/activity_point.rb (hunk 8 above)
Comment (alkeshrails) [reply app/models/activity_point.rb]: Done

File: app/models/leaderboard/base_leaderboard.rb (hunk 5 above)
Comment (alkeshrails) [reply app/models/leaderboard/base_leaderboard.rb:6]: Done

File: db/migrate/20200709123607_add_ml_challenge_id_to_submission.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (alkeshrails) [reply db/migrate/20200709123607_add_ml_challenge_id_to_submission.rb:3]: Done

File: db/migrate/20200715185739_create_participant_ml_challenge_goals.rb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+class CreateParticipantMlChallengeGoals < ActiveRecord::Migration[5.2]
+  def change
+    create_table :participant_ml_challenge_goals do |t|
```
[the remark this answers is not in the document]
Comment (alkeshrails) [reply db/migrate/20200715185739_create_participant_ml_challenge_goals.rb:3]: Done

File: db/migrate/20200721161118_add_ml_challenge_id_to_base_leaderboard.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class AddMlChallengeIdToBaseLeaderboard < ActiveRecord::Migration[5.2]
+  def change
+    add_column :base_leaderboards, :ml_challenge_id, :integer
```
[the remark this answers is not in the document]
Comment (alkeshrails) [reply db/migrate/20200721161118_add_ml_challenge_id_to_base_leaderboard.rb:3]: added index

**Pull request merged**